gwr-spotter/src/app/tests.rs

```diff
@@ -0,0 +1,410 @@
+// Copyright (c) 2025 Graphcore Ltd. All rights reserved.
+
+#[cfg(test)]
+mod tests {
+    // Note that all tests need to be run serially as spotter uses a global state
+    // shared between threads.
+
+    use std::sync::mpsc::{Receiver, channel};
+    use std::sync::{Arc, Mutex};
+
+    use serial_test::serial;
+
+    use crate::app::{App, InputState};
+    use crate::filter::Filter;
+    use crate::renderer::Renderer;
+    use crate::rocket::SHARED_STATE;
+
+    const FRAME_HEIGHT: usize = 30;
+    const NUM_RENDER_LINES: usize = 200;
+
+    /// Create a "mock" renderer: use the real `Renderer` but just pre-define
+    /// minimal state so we can test movement behaviour.
+    fn create_mock_renderer() -> Renderer {
+        let mut renderer = Renderer::new();
+        // Set a reasonable frame size so block moves are non-zero.
+        renderer.set_frame_size(FRAME_HEIGHT);
+
+        // Pretend we have renderable lines so movement code has
+        // something to clamp against.
+        renderer.num_render_lines = NUM_RENDER_LINES;
+        renderer
+    }
+
+    /// Create a "mock" filter: can use an actual `Filter`, but using a channel
```

Comment:
I wasn't quite sure where the typo was in this sentence, though I think there is one. Is it supposed to be `but` -> `by`?